packages/muelu/src/Graph/MatrixTransformation/MueLu_MatrixConstruction.hpp

```diff
@@ -907,6 +909,209 @@ class VectorFillFunctor {
   }
 };
 
+template <class local_matrix_type>
+class MergeCountFunctor {
+ private:
+  using scalar_type             = typename local_matrix_type::value_type;
+  using local_ordinal_type      = typename local_matrix_type::ordinal_type;
+  using memory_space            = typename local_matrix_type::memory_space;
+  using results_view            = Kokkos::View<DecisionType*, memory_space>;
+  using block_indices_view_type = Kokkos::View<local_ordinal_type*, memory_space>;
+  using permutation_type        = Kokkos::View<local_ordinal_type*, memory_space>;
+
+  using rowptr_type = typename local_matrix_type::row_map_type::non_const_type;
+  using ATS         = Kokkos::ArithTraits<local_ordinal_type>;
+
+  local_matrix_type A;
+  local_ordinal_type blockSize;
+  block_indices_view_type ghosted_point_to_block;
+  rowptr_type merged_rowptr;
+
+  BlockRowComparison<local_matrix_type> comparison;
+  permutation_type permutation;
+
+ public:
+  MergeCountFunctor(local_matrix_type& A_, local_ordinal_type blockSize_, block_indices_view_type ghosted_point_to_block_, rowptr_type& merged_rowptr_)
+    : A(A_)
+    , blockSize(blockSize_)
+    , ghosted_point_to_block(ghosted_point_to_block_)
+    , merged_rowptr(merged_rowptr_)
+    , comparison(BlockRowComparison(A, blockSize_, ghosted_point_to_block)) {
+    permutation = permutation_type("permutation", A.nnz());
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const local_ordinal_type brlid, local_ordinal_type& nnz_graph, const bool& final) const {
```

Comment:
Since this may be called more than once, is it worth trying to avoid the potential extra `Misc::serialHeapSort` call, e.g.?

Reply:
I'm not so sure about that. Wouldn't that mean that either:
- we permute the input matrix, or
- we trade computation for additionial memory?